src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,401 @@
+  struct SerializedResult {
+    kj::Array<kj::byte> data;
+    bool caughtException;
+  };
+
+  // Serializes the return value of our method.
+  SerializedResult serialize(jsg::Lock& js, jsg::JsValue value) {
+    kj::Maybe<SerializedResult> result;
+
+    js.tryCatch([&] {
+      auto data = serializeV8(js, value.addRef(js).getHandle(js));
+      // Serialized okay!
+      result = SerializedResult { .data = kj::mv(data), .caughtException = false };
+    }, [this, &js, &result](jsg::Value exception) {
+      auto fullError = kj::str(jsg::check(exception.getHandle(js)->ToString(js.v8Context())));
+
+      // We may need to slice part of the message so we don't have `Error: Error: msg`.
+      kj::Maybe<kj::StringPtr> errorMessage(fullError);
+      if (fullError.startsWith("Error: "_kj)) {
+        errorMessage = fullError.slice(kj::size("Error: "_kj));
+      }
+
+      // Failed to serialize data.
+      auto failedResult = serialize(js, js.error(KJ_REQUIRE_NONNULL(errorMessage)));
+      // We are in the error branch, so we must set caughtException.
+      failedResult.caughtException = true;
+      result = kj::mv(failedResult);
+    });
+    return kj::mv(KJ_REQUIRE_NONNULL(result));
+  }
+
+  // Invokes the given function and serializes the result for RPC.
+  kj::Promise<SerializedResult> callFuncAndSerialize(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    // `processAsyncResult` expects a promise, so even if we return synchronously we want to wrap
+    // the result of calling the JS function in a promise.
+    auto wrapValInPromise = [](jsg::Lock& js, v8::Local<v8::Value> value) {
+      auto resolver = jsg::check(v8::Promise::Resolver::New(js.v8Context()));
+      auto promise = resolver->GetPromise();
+
+      jsg::check(resolver->Resolve(js.v8Context(), value));
+      return js.toPromise(kj::mv(promise));
+    };
+
+    kj::Maybe<jsg::Promise<jsg::Value>> eventualResult;
+    bool caughtException = false;
+    js.tryCatch([&] {
+      eventualResult = wrapValInPromise(js, invokeFn(js, fn, serializedArgs));
+    }, [&](jsg::Value exception) {
+      // We threw while executing JS.
+      caughtException = true;
+      eventualResult = wrapValInPromise(js, exception.getHandle(js));
+    });
+
+    auto& result = KJ_REQUIRE_NONNULL(eventualResult);
+    auto serializedResult = co_await processAsyncResult(js, kj::mv(result));
+    if (caughtException) {
+      serializedResult.caughtException = caughtException;
+    }
+    co_return kj::mv(serializedResult);
   }
+
+  // The JS method returned a promise and we need to wait for the promise to resolve or reject
+  // before serializing the result.
+  kj::Promise<SerializedResult> processAsyncResult(jsg::Lock& js, jsg::Promise<jsg::Value> promise) {
+    auto& ctx = IoContext::current();
+    co_return co_await ctx.awaitJs(js, promise
```

Comment:
The pattern `co_await ctx.awaitJs(` implies that we're dropping the isolate lock and then taking it again when we don't really have to. That is: `ctx.awaitJs()` converts a JS promise (which normally runs on the V8 mircotask loop, which is pumped with the isolate lock held continuously) into a KJ promise (which runs on the KJ event loop, which can only be pumped after we've released all locks).

If you instead use `promise.then()` here, you can stay entirely in JS land, using the V8 microtask loop, so you don't have to drop to KJ and unnecessarily release the isolate lock. You would also want to make this function return `jsg::Promise` instead of `kj::Promise`. Again, unfortunately, we don't support coroutines with jsg::Promise yet, so you have to use the old `.then()` syntax.